### ISSUE TITLE
Support for DSD in Wavpack

### ISSUE THREAD

**Issue by terual** [LMS-Community/slimserver#424]
The current Wavpack version used is 4.50.1 while DSD support for Wavpack was introduced in version 5.0.0 (see http://www.wavpack.com/changelog.txt).

When we update Wavpack to version 5 (latest is 5.3.0) it is a matter of simply adding the following to custom-convert.conf to use Wavpack for DSD-files (Note: it is then not possible anymore to use Wavpack for PCM).

Without DoP (converting DSD to PCM):
```
wvp flc * *
    # FT:{START=--skip=%t}U:{END=--until=%v}D:{RESAMPLE=-r %d}
    [wvunpack] $FILE$ --dff $START$ $END$ -o - | [dsdplay] $RESAMPLE$
```

DoP enabled (not tested yet):
```
wvp flc * *
    # FT:{START=--skip=%t}U:{END=--until=%v}D:{RESAMPLE=-r %d}
    [wvunpack] $FILE$ --dff $START$ $END$ -o - | [dsdplay] -u $RESAMPLE$
```

**Comment by terual** [comment]
See https://github.com/Logitech/slimserver-vendor/pull/79 for the changes for the *nix binaries. 

I do not know how Windows binaries are built, but maybe the binaries release by WavPack can be used (https://github.com/dbry/WavPack/releases/download/5.3.0/wavpack-5.3.0-x86.zip).

**Comment by mherger** [comment]
Thanks! slimserver-vendor indeed is the right place for binary related changes.

Alas I'm wondering about the limitation you mentioned: you say it was not possible to use "Wavpack for PCM". IMHO that's a killer. I bet there are many more users using PCM than DSD, right? Is the reason the `--dff` parameter added to the transcoding command? In this case it would probably be smarter to treat those files differently. Any way to distinguish a PCM from a DSD file at scan time?

All that said I believe it would be great if you could wrap all this up in a plugin. I don't see me break PCM playback in favour of DSD in core LMS.

**Comment by terual** [comment]
Of course PCM playback is more important than DSD playback for most users. What I am proposing is just updating the WavPack binaries so that users which want to use WavPack for DSD-compression can add the custom-convert.conf changes. There are no further changes neccesary. But it may be more user friendly to wrap those custom-convert.conf rules into a plugin. If you then enable the plugin, WavPack is used for DSD and otherwise for PCM.

By the way, Audio::Scan is able to distinguish between DSD and PCM I think. Would it then be possible to use WavPack for both DSD as PCM?

**Comment by terual** [comment]
I have published a plugin which enables the WavPack DSD convert rules based on the configuration from the DSDPlayer plugin. This seems to work for me with a recompiled wvunpack (version 5.3.0).

https://github.com/terual/DSDWavPack (link doesn't work, I've added my changes to DSDPlayer plugin, see below)

**Comment by mherger** [comment]
There's no way to set the necessary pref per client, is there?

**Comment by terual** [comment]
I think that is exactly what the plugin does, or do I not understand your comment? Or do you mean that the plugin should be configurable per client?

**Comment by mherger** [comment]
It does read the pref, but doesn't set it anywhere, nor does it provide a means for the user to set it?

**Comment by terual** [comment]
It reads the prefs (DoP and resampling choices) from the DSDPlayer plugin and subsequently adds the convert rules for all clients. So configuration must take place in the DSDPlayer plugin. It also uses the dsdplay binary from that plugin. 

Also, it does not make sense to make this configurable per client, because then DSD Wavpacks can only be played on the client for which you have configured this and not on others. And for PCM Wavpacks the other way around.

**Comment by mherger** [comment]
Oh, I didn't realise this was adding/inheriting from the DSDPlayer plugin. Would it make sense to fork it?

**Comment by terual** [comment]
I guess a fork with an extra configuration where you can choose between WavPack for DSD or WavPack PCM would make sense. When I have some time I will work on it.

In the meantime, all this functionality relies on the upgrade of WavPack to version 5. What steps need to be taken to get this in LMS?

**Comment by terual** [comment]
By the way, do you know the upstream of http://triodeplugins.herger.net/DSDPlayer-linux-v1.0.1.zip and http://triodeplugins.herger.net/DSDPlayer-windows-v1.0.1.zip? Instead of a fork, upstreaming would be better I think.

**Comment by michaelherger** [comment]
Triode unfortunately disappeared from the SB land a few years ago. When Google shut down their repositories I copied some of Triode's archives and am offering them for download on my server. But the other guy mentioned in the plugin is Kimmo Taskinen, who's still around. I shot him a mail asking about this.

**Comment by terual** [comment]
I found the DSDPlayer plugin in Ralph Irvings account on GitHub (https://github.com/ralph-irving/DSDPlayer) and added my changes to support DSD WavPacks on top of that: https://github.com/terual/DSDPlayer/commit/afa7755d97d9a2893950d8bb0244c612a147e62e

**Comment by kimmot** [comment]
We discussed this by email with Michael and he asked me to comment here (sorry I didn't read everything above here). Here is a quick patch for LMS that should classify WavPack DSD files as "wvpx" type and the rest WavPack files would remain as "wvp". This would enable the separate conversion rules for WavPack DSD files.

[lms_wavpack_dsd_type_patch.txt](https://github.com/Logitech/slimserver/files/5283048/lms_wavpack_dsd_type_patch.txt) see version 2 below.

**Comment by terual** [comment]
Thanks, that would be the best solution! As soon as the WavPack binaries have been updated and this patch is applied I can add the necessary convert rules to the DSDPlayer plugin. The question remains if the DSDPlayer plugin is still being maintained. @kimmot can you shed any light on that?

**Comment by kimmot** [comment]
Here is version 2 for the patch. Better to detect from "bits_per_sample" and WVPX added to strings.txt.
[lms_wavpack_dsd_type_patch_v2.txt](https://github.com/Logitech/slimserver/files/5283329/lms_wavpack_dsd_type_patch_v2.txt)

**Comment by kimmot** [comment]
I haven't ever maintained or developed DSDPlayer-plugin. The history is such that I originally developed PlayDSD plugin for Daphile and shared it to Triode. He made some modifications and published it as a generic DSDPlayer plugin. Afterwards I have changed the functionality in Daphile (generic transcoder) and it does not even use the original binary (dsdplay?; Daphile uses sox from mansr's github repo).

**Comment by ralph-irving** [comment]
In response to a email from terual.
* My DSDPlayer repository contains additional dsdplay binaries I've built for arm and armhf linux, freebsd x64, macos and windows.  I have not added my dsdplayer 1.05 zip to any repository at this point.  I see you've forked it and added the wavplay dsd changes as v1.06.  Feel free to have your version published to a repo to replace the current v1.02.

**Comment by terual** [comment]
Sounds like a plan! 

@michaelherger Is there a way to publish the plugin to be automatically picked up by https://repos.squeezecommunity.org/extensions.xml? I saw a public.xml in the lms-material repo (https://github.com/CDrummond/lms-material/blob/master/public.xml) which seems to be used for the extensions.xml on your https://repos.squeezecommunity.org site.

**Comment by michaelherger** [comment]
Just give me the URL of the repository file, and I'll include it in the main repository.

**Comment by terual** [comment]
I've created a pull request based on Kimmo's patch: https://github.com/Logitech/slimserver/pull/427
The repository file for the DSDPlayer plugin is here: https://github.com/terual/DSDPlayer/blob/master/repo.xml

**Comment by michaelherger** [comment]
Thanks @terual! I see that you provide one plugin version for Unix/Linux only. Is this correct? Other platforms won't be able to take advantage of this change?

**Comment by terual** [comment]
Good catch! Oversight on my part, as the plugin is multiplatform and binaries are supplied for all platforms. I have changed the repo.xml

**Comment by mherger** [comment]
Great! I updated the repository file accordingly. Thanks!

**Comment by terual** [comment]
Thanks! Last part of the DSD WavPack support is the upgrade of the WavPack binary (see https://github.com/Logitech/slimserver-vendor/pull/79). Is it a manual process of building the binaries in slimserver-vendor and copying them to this repository? Anything I can do?

**Comment by michaelherger** [comment]
Yes, that would be it. But I must admit that I usually hesitate to replace binaries lightly... But you could include them in the plugin? Would that work?

**Comment by terual** [comment]
It seems that if I include the wvunpack binary in the plugin, it doesn't take preference to the binary supplied by LMS.

On the other hand, WavPack supplies binaries for mac, win32 and win64 (http://www.wavpack.com/downloads.html). Only for *nix binaries need to be built. Maybe that makes the process a bit more manageable.

**Comment by michaelherger** [comment]
Ok, added updated binaries for the most important platforms. Verified the ARM binary on Raspberry Pi OS and piCorePlayer. That will hopefully cover the vast majority of installations.

**Comment by terual** [comment]
Thank you very much!

**Comment by kimmot** [comment]
Now that there's wvunpack version >5 in LMS, you should merge attached change to get generic conversion rule from wvpx to dff or dsf. Then Squeezelite with DSD support would not even need any extra plugin as it supports DoP (or native DSD, Linux only?) output.
[convert_wavpack_dsd_to_dsf_or_dff.txt](https://github.com/Logitech/slimserver/files/5348246/convert_wavpack_dsd_to_dsf_or_dff.txt)

**Comment by terual** [comment]
Good idea! I have created the following commit with your changes: https://github.com/terual/slimserver/commit/7e2c0b9b4af39f1058dccd5e131454f1e87f55c1

I have ommited the D in the capabilities for the wvpx->dsf/dff case. And I also included a wvpx->flc convert rule, because wvunpack can also convert to pcm if you use the -w flag. Does that look good to you?

**Comment by kimmot** [comment]
You should have not dropped D. It won't work without it because none of the players report to support DSD sample rates and that's why LMS won't ever use the conversion rule without D (I tested it first without D).

I would be very careful on implementing the "wvpx->flc" conversion rule. That could be a problem for the cases where user wants to output DoP encoded DSD with the help of DSDPlayer plugin. I'm not sure, maybe the player specific conversion provided by the plugin will override.

**Comment by terual** [comment]
The DSDPlayer plugin adds conversion rules specific for players which override the generic rules. And I've added the D again: https://github.com/terual/slimserver/commit/cf7eadcbf3ad2e983853164aabf45cd1edf59c32

Only thing I still want to test is if the generic wvpx->dsf/dff rule gets precendence to the generic wvpx->flc rule for squeezelite players.

**Comment by terual** [comment]
I have done some testing with squeezelite v1.8:
squeezelite without options or with the '-D' option (DoP) picks the wvpx->dsf rule.
squeezelite with option '-c flac' or '-e dsd' picks the wvpx->flc rule.

So I think it looks good. Apparently the dsf rule gets precedence except when squeezelite does not support DSD.

**Comment by terual** [comment]
I've made a pull request: https://github.com/Logitech/slimserver/pull/435

**Comment by ralph-irving** [comment]
I've built wvunpack 5.3.0 for linux i386 and arm.
https://sourceforge.net/projects/lmsclients/files/utility/wvunpack-5.3.0-armel.tar.gz/download
https://sourceforge.net/projects/lmsclients/files/utility/wvunpack-5.3.0-i386.tar.gz/download

**Comment by michaelherger** [comment]
Thanks @ralph-irving! Any idea why your binaries would be half the size from before, and much smaller than other builds?

**Comment by ralph-irving** [comment]
The i386 build was on centos6 which I use for all my i386 binaries.  The current i386 wvunpack 4.5 was built with the linux 2.4 kernel.  Could be the optimizations are much better now.

The arm binaries are different eabi versions 4 vs 5.  I don't know if that makes any difference to the size

**wvunpack**: ELF 32-bit LSB executable, ARM, **EABI5** version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.3, for GNU/Linux 2.6.18, BuildID[sha1]=8d403208c29b7fca03f5ca8acd9b14657408c54e, stripped

**wvunpack-4.5**: ELF 32-bit LSB executable, ARM, **EABI4** version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.3, for GNU/Linux 2.6.14, stripped

I did notice that my armel binary was linked with the shared libgcc_s library, so I rebuilt it linking against the static version but that only increased the binary size by 7k.  The newer build is now on sf, same link.
